flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java

```diff
@@ -0,0 +1,182 @@
+					zis.putNextEntry(zipEntry);
+
+					IOUtils.copyBytes(new ByteArrayInputStream(testFileContent.getBytes(StandardCharsets.UTF_8)), zis, false);
+				}
+				return zipArchive;
+			} else {
+				throw new IllegalArgumentException("This service contains only entry for " + permanentBlobKey);
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+
+		}
+	};
+
+	@Before
+	public void setup() throws Exception {
+		fileCache = new FileCache(new String[]{temporaryFolder.newFolder().getAbsolutePath()}, blobService);
+	}
+
+	@After
+	public void shutdown() {
+		fileCache.shutdown();
+	}
+
+	@Test
+	public void testDirectoryDownloadedFromBlob() throws Exception {
+		JobID jobID = new JobID();
+		ExecutionAttemptID attemptID = new ExecutionAttemptID();
+
+		final String fileName = "test_file";
+		// copy / create the file
+		final DistributedCache.DistributedCacheEntry entry = new DistributedCache.DistributedCacheEntry(
+			fileName,
+			false,
+			InstantiationUtil.serializeObject(permanentBlobKey),
+			true);
+		Future<Path> copyResult = fileCache.createTmpFile(fileName, entry, jobID, attemptID);
+
+		final Path dstPath = copyResult.get();
+		final FileSystem fs = dstPath.getFileSystem();
+		final FileStatus fileStatus = fs.getFileStatus(dstPath);
+		assertTrue(fileStatus.isDir());
+
+		final Path cacheFile = new Path(dstPath, "cacheFile");
+		assertTrue(fs.exists(cacheFile));
+		final String actualContent = FileUtils.readFileUtf8(new File(cacheFile.getPath()));
+		assertEquals(testFileContent, actualContent);
+	}
+
+	@Test
+	public void testDirectoryCleanUp() throws Exception {
+		JobID jobID = new JobID();
+		ExecutionAttemptID attemptID1 = new ExecutionAttemptID();
+		ExecutionAttemptID attemptID2 = new ExecutionAttemptID();
+
+		final String fileName = "test_file";
+		// copy / create the file
+		final DistributedCache.DistributedCacheEntry entry = new DistributedCache.DistributedCacheEntry(
+			fileName,
+			false,
+			InstantiationUtil.serializeObject(permanentBlobKey),
+			true);
+		Future<Path> copyResult = fileCache.createTmpFile(fileName, entry, jobID, attemptID1);
+		fileCache.createTmpFile(fileName, entry, jobID, attemptID2);
+
+		final Path dstPath = copyResult.get();
+		final FileSystem fs = dstPath.getFileSystem();
+		final FileStatus fileStatus = fs.getFileStatus(dstPath);
+		final Path cacheFile = new Path(dstPath, "cacheFile");
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		fileCache.releaseJob(jobID, attemptID1);
+		// still should be available
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		fileCache.releaseJob(jobID, attemptID2);
+		// still should be available, file will be deleted after 5 seconds
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		// after a while, the file should disappear
+		long deadline = System.currentTimeMillis() + 10000;
+		do {
+			Thread.sleep(5500);
```

Comment:
well, I'd rather not blow 5-10 seconds on a single test.

I would suggest shutting down the FileCache instead. If the delete process was triggered it should still delete the file.

Reply:
I can not just use `fileCache.shutdown` because the shutdown performs additional clearing of temporary directories, which is not what I want to test. I've changed the implementation. I am not sure thought about making the `ExecutorService` `VisibleForTesting`. Another alternative would be to revert to a similar loop, but this time with a shorter period.